packages/url-loader/src/lib/upload-widget.ts

```diff
@@ -0,0 +1,114 @@
+  // https://cloudinary.com/documentation/upload_widget#signed_uploads
+
+  const { cloudName, apiKey } = config?.cloud || {};
+
+  if (!cloudName) {
+    throw new Error('A Cloudinary Cloud name is required, please make sure your environment variable is set and configured in your environment.');
+  }
+
+  if (signed && !apiKey) {
+    throw new Error('A Cloudinary API Key is required for signed requests, please make sure your environment variable is set and configured in your environment.');
+  }
+  
+  if ( !signed && !options.uploadPreset ) {
+    throw new Error('A Cloudinary Upload Preset is required for unsigned uploads. Please specify an uploadPreset or configure signed uploads.');
+  }
+
+  const uploadOptions: CloudinaryUploadWidgetOptions = {
+    cloudName,
+    apiKey,
+    ...options,
+  };
+
+  if ( signed ) {
+    uploadOptions.uploadSignature = uploadSignature;
+  }
+
+  return uploadOptions;
+}
+
+
+/**
+ * generateUploadWidgetResultCallback
+ */
+
+export type CloudinaryUploadWidgetResultCallback = (results: CloudinaryUploadWidgetResults) => void;
+export type CloudinaryUploadWidgetErrorCallback = (error: CloudinaryUploadWidgetError, results: CloudinaryUploadWidgetResults) => void;
+
+export interface GenerateUploadWidgetResultCallback {
+  onOpen?: CloudinaryUploadWidgetResultCallback;
+  /**
+   * @deprecated use onSuccess instead
+   */
+  onUpload?: CloudinaryUploadWidgetResultCallback;
+  onAbort?: CloudinaryUploadWidgetResultCallback;
+  onBatchCancelled?: CloudinaryUploadWidgetResultCallback;
+  onClose?: CloudinaryUploadWidgetResultCallback;
+  onDisplayChanged?: CloudinaryUploadWidgetResultCallback;
+  onPublicId?: CloudinaryUploadWidgetResultCallback;
+  onQueuesEnd?: CloudinaryUploadWidgetResultCallback;
+  onQueuesStart?: CloudinaryUploadWidgetResultCallback;
+  onRetry?: CloudinaryUploadWidgetResultCallback;
+  onShowCompleted?: CloudinaryUploadWidgetResultCallback;
+  onSourceChanged?: CloudinaryUploadWidgetResultCallback;
+  onSuccess?: CloudinaryUploadWidgetResultCallback;
+  onTags?: CloudinaryUploadWidgetResultCallback;
+  onUploadAdded?: CloudinaryUploadWidgetResultCallback;
+  onError: CloudinaryUploadWidgetErrorCallback;
+  onResult: CloudinaryUploadWidgetResultCallback;
+}
+
+const UPLOAD_WIDGET_EVENTS: { [key: string]: string } = {
+  'abort': 'onAbort',
+  'batch-cancelled': 'onBatchCancelled',
+  'close': 'onClose',
+  'display-changed': 'onDisplayChanged',
+  'publicid': 'onPublicId',
+  'queues-end': 'onQueuesEnd',
+  'queues-start': 'onQueuesStart',
+  'retry': 'onRetry',
+  'show-completed': 'onShowCompleted',
+  'source-changed': 'onSourceChanged',
+  'success': 'onSuccess',
+  'tags': 'onTags',
+  'upload-added': 'onUploadAdded',
+}
+
+export function generateUploadWidgetResultCallback(options: GenerateUploadWidgetResultCallback) {
+  return function resultCallback(error: CloudinaryUploadWidgetError, uploadResult: CloudinaryUploadWidgetResults) {
+    if ( error && error !== null ) {
```

Comment:
question: is this full statement necessary? AFAIK, null is falsy so if error is truthy then it means automatically that it cannot be null (falsy)

Reply:
null is falsy yea

Reply:
sounds good, was just pasting this over from existing code, will clean that up!